### PR TITLE
osd/ReplicatedBackend: handle tailing zeros of recovering objects nicer

### PR DESCRIPTION
**Pull request by Songweibin**
E.g.,
(1) recovery_info.copy_subset = [0, 4194304]
(2) copy_subset(fiemap) = [0, 8192]
(3) out_op->data_included.span_of(copy_subset, 0, 4194304) = [0, 8192]

We'll set new_progress.data_recovered_to = 8192 as a result, and
fail to complete the push in one go.

Signed-off-by: xie xingguo <xie.xingguo@zte.com.cn>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/osd/ReplicatedBackend.cc [hunk 1]
```diff
@@ -1961,7 +1961,8 @@ int ReplicatedBackend::build_push_op(const ObjectRecoveryInfo &recovery_info,
 
       out_op->data_included.span_of(copy_subset, progress.data_recovered_to,
                                     available);
-      if (out_op->data_included.empty()) // zero filled section, skip to end!
+      if (out_op->data_included.empty() ||
+          out_op->data_included == copy_subset) // zero filled section, skip to end!
```
Comment (tchaikov) [inline src/osd/ReplicatedBackend.cc:1965]:
```suggestion
          out_op->data_included.range_end() == copy_subset.range_end()) // zero filled section, skip to end!
```

Comment (Songweibin) [reply src/osd/ReplicatedBackend.cc]: @tchaikov  Updated, thank you.

**Review by tchaikov — APPROVED** [review]

**Review by jdurgin — APPROVED** [review]
nice catch!

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-12-23_05:37:18-rados-wip-kefu-testing-2020-12-23-1139-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-12-24_04:12:19-rados-wip-kefu-testing-2020-12-23-1139-distro-basic-smithi/

failures tracked by

- https://tracker.ceph.com/issues/48323
- https://tracker.ceph.com/issues/48712

**Pull request merged**